### PR TITLE
Save extra AOVs

### PR DESCRIPTION
**Pull request by oktomus**
We are now able to:
- Successfully save extra AVOS as exr
- Save extra AOVs as png if possible

Fix for #1858 

I am not sure if you are ok regarding changes on ` Frame::create_extra_aov_image`.
I haven't been able to test futher with png because I have a linking issue (possible version mismatch) which cause that `pngptr` is null in `PNGImageFileWriter::write`.

I was also thinking of a different approach:
Create new AVO classes for data images and color images. Is this possible ?

### CODE REVIEW

**Comment by est77** [comment]
Thanks for the PR!
I'll check it today after work.

**Comment by dictoon** [comment]
Thanks! Small remark: it's spelled AOV for Arbitrary Output Variables, not AVO.

**Comment by oktomus** [comment]
Wooups, I know and still I made the mistake a lot

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 1]
```diff
@@ -776,6 +804,30 @@ bool Frame::write_main_and_aov_images() const
...
 
+    if (!aov_images().empty())
+    {
+        const bf::path boost_file_path(filepath);
+        const bf::path directory = boost_file_path.parent_path();
+        const string base_file_name = boost_file_path.stem().string();
+        const string extension = boost_file_path.extension().string();
+
+        // Extra AOVs
+        for (size_t i = 0, e = aov_images().size(); i < e; ++i)
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: This fragment of code is repeated. Refactor into a function in an unnamed namespace if possible, 
if not, refactor to a private method of the frame.

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 2]
```diff
@@ -91,8 +91,8 @@ namespace
         {
             if (m_params.m_diagnostics)
             {
-                m_variation_aov_index = frame.create_extra_aov_image("variation");
-                m_samples_aov_index = frame.create_extra_aov_image("samples");
+                m_variation_aov_index = frame.create_extra_aov_image("variation", PixelFormatHalf);
```
Comment (est77) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: With these changes, the range and precision of the images used by the adaptive pixel renderer is much lower than before. This is going to affect the results of the adaptive sampling. I think we want to keep them as floats.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 3]
```diff
@@ -278,11 +278,13 @@ const AOVContainer& Frame::internal_aovs() const
     return impl->m_internal_aovs;
 }
 
-size_t Frame::create_extra_aov_image(const char* name) const
+size_t Frame::create_extra_aov_image(
+    const char* name,
+    const PixelFormat pixel_format) const
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Personally, I prefer to undo this change and remove the option of saving extra aov images to pngs.
The images in the extra aov images are always non-color images and in my opinion, it doesn't make much sense to save them as pngs.

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp (hunk 2 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: I chanded the format because the tile_end maps values to colors.

**Comment by oktomus** [comment]
Thanks for the review @est77, I will commit asap :)

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 4]
```diff
@@ -720,7 +765,7 @@ bool Frame::write_aov_images(const char* file_path) const
 
     bool success = true;
 
-    if (!aovs().empty())
+    if (!aovs().empty() || !aov_images().empty())
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Remove the !aov_images().empty() check

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 5]
```diff
@@ -687,6 +700,38 @@ namespace
 
         return true;
     }
+
+    bool write_extra_aovs(
+        const ImageStack & images,
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Remove space between the types and the '&' and align parameter names vertically.

**Comment by est77** [comment]
Thanks for the changes.

There is a problem that I missed yesterday.
In write_extra_aovs(), we save all the images in the image stack:

for (size_t i = 0, e = images.size(); i < e; ++i)

but many AOVs also create images on the image stack.
As a result, we will save them twice, once when the AOVs are written and a second time in write_extra_aovs.

Maybe the easiest way to avoid saving some images twice is to keep a vector of the image indices created by Frame::create_extra_aov_image and in write_extra_aovs, write only those images.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 6]
```diff
@@ -646,6 +649,16 @@ namespace
             }
             else if (extension == ".png")
             {
+                // Test if the image contains appropriate data
```
[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Yes, I forgot that

File: src/appleseed/renderer/modeling/frame/frame.cpp (hunk 4 above)
Comment (oktomus) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Can you explain me why ?

**Comment by oktomus** [comment]
Thanks for the review, I will do that :)

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 7]
```diff
@@ -720,27 +763,34 @@ bool Frame::write_aov_images(const char* file_path) const
 
     bool success = true;
 
-    if (!aovs().empty())
```
Comment (oktomus) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: I do not check for the size anymore because if I do, I would have to check `aovs().empty()` and `aov_images().empty()`. Otherwise extra AOVs wouldn't be saved if there isn't any main AOV.

Comment (oktomus) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Maybe I'm wrong, there is always an AOV, beauty, right ? @est77

Comment (est77) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: beauty is not an AOV in appleseed. There's an image in the frame to store it.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 8]
```diff
@@ -114,13 +114,15 @@ struct Frame::Impl
...
     DenoisingMode           m_denoising_mode;
+    bool                    m_save_extra_aovs;
 
     // Images.
     unique_ptr<Image>       m_image;
     unique_ptr<ImageStack>  m_aov_images;
     AOVContainer            m_aovs;
     AOVContainer            m_internal_aovs;
     DenoiserAOV*            m_denoiser_aov;
+	vector<size_t>          m_extra_aovs;
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: align the type and the var member name with the previous lines.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 9]
```diff
@@ -280,9 +284,12 @@ const AOVContainer& Frame::internal_aovs() const
 
 size_t Frame::create_extra_aov_image(const char* name) const
 {
-    const size_t index = aov_images().get_index(name);
+    size_t index = aov_images().get_index(name);
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: index can be const. We don't modify it in the method.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 10]
```diff
@@ -687,6 +694,42 @@ namespace
 
         return true;
     }
+
+    bool write_extra_aovs(
+        const ImageStack&       images,
+            const vector<size_t>&   aov_indices,
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: align with the rest of the arguments.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 11]
```diff
@@ -687,6 +694,42 @@ namespace
...
+
+        if (extension != ".exr")
+        {
+            RENDERER_LOG_ERROR("extra AOVs can only be saved as exr.");
+            return false;
+        }
+
+        for (size_t i = 0, e = aov_indices.size(); i < e; ++i)
+        {
+            size_t image_index = aov_indices[i];
```
Comment (est77) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: make const

**Comment by est77** [comment]
Thanks for the changes!
Only a few very small details and we can merge.

**Comment by oktomus** [comment]
Thanks for the review, I'm really sorry for all these align issues.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 12]
```diff
@@ -720,27 +764,34 @@ bool Frame::write_aov_images(const char* file_path) const
...
+        // Write AOV image.
+        if (!write_image(aov_file_path.c_str(), aov->get_image(), aov))
+            success = false;
+    }
 
-            // Write AOV image.
-            if (!write_image(aov_file_path.c_str(), aov->get_image(), aov))
-                success = false;
-        }
+    if (impl->m_save_extra_aovs && !aov_images().empty())
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Why checking `!aov_images().empty()`? Isn't this test redundant?

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 13]
```diff
@@ -687,6 +695,42 @@ namespace
...
+    bool write_extra_aovs(
+        const ImageStack&       images,
+        const vector<size_t>&   aov_indices,
+        const bf::path&         directory,
+        const string&           base_file_name,
+        const string&           extension)
+    {
+        bool success = true;
+
+        if (extension != ".exr")
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Is `extension` guaranteed to always be a lower case?

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 14]
```diff
@@ -821,6 +888,21 @@ void Frame::write_main_and_aov_images_to_multipart_exr(const char* file_path) co
...
     }
 
+    if (impl->m_save_extra_aovs)
+    {
+        for (size_t i = 0, e = impl->m_extra_aovs.size(); i < e; ++i)
+        {
+            const size_t image_index = impl->m_extra_aovs[i];
+            assert(image_index < aov_images().size());
+
+            const Image & image = aov_images().get_image(image_index);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Remove space before `&`.

**Comment by dictoon** [comment]
Very nice work @oktomus! Appreciated.

**Comment by dictoon** [comment]
@est77 When merging this, make sure to squash the PR into a single commit, and to remove the individual commit titles from the commit body (which GitHub fills automatically).

**Comment by est77** [comment]
Done!
